### PR TITLE
TPP 00064

### LINKED ISSUE #64: TPP-00064: Calendar Service AsyncStorage Endpoints
> As a user...I want the inputs I'm providing the front-end to be saved in a backend, and I want to read from those saved values. 

API documentation: https://github.com/uoftblueprint/the-period-purse/wiki/Calendar-API

In a file titled CalendarService.js, implement the APIs specified in the API documentation that the Month View Calendar will need. You can check that these are actually being stored/check what's in AsyncStorage by installing the React Native Debugger, instructions found in the README. 

Example of an API Service file Faye made: https://github.com/uoftblueprint/the-period-purse/wiki/Example-API-Service-File

Official AsyncStorage docs: https://react-native-async-storage.github.io/async-storage/docs/usage

for .catch(error) just do console.log(error); for now! We'll have a story go through these later.

You should be importing it like this: 
```import AsyncStorage from '@react-native-async-storage/async-storage';```
DO NOT USE 
```import AsyncStorage from react-native;```

### PR DESCRIPTION
**Pull request by edwardhuahan**
# Description

Implemented the API required for Calendar Service specified in the [wiki](https://github.com/uoftblueprint/the-period-purse/wiki/Models-and-AsyncStorage-Keys#calendar)

Fixes #64 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if required)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tpp-app/src/services/CalendarService.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import { KEYS, TRACK_SYMPTOMS } from './utils/constants.js'
+import { Symptoms } from './utils/models.js';
+
+/**
+ * Makes API request to AsyncStorage to access symptoms for the entire year
+ * @param year: number
```
Comment (kenneth-miura) [inline tpp-app/src/services/CalendarService.js]: Take a look at this [link](https://jsdoc.app/tags-param.html). Basically u can do type notation on an argument like `@param {number} year DESCRIPTION`. Makes it show up when u mouse over the signature in VS code & prob other editors. very cool, very hip

File: tpp-app/src/services/CalendarService.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * @returns: object
+ */
+export const GETYearData = async (year) => {
```
Comment (kenneth-miura) [inline tpp-app/src/services/CalendarService.js:11]: Think there is some overlap with the function `getCalendarByYear` I wrote in my PR, but I think it's fine. There's a difference, namely that the method I wrote gets the current calendar, & the prev & next year's calendar if it exists. It's currently not converting all the items to symptoms, but I can swap it so it calls this function to make functionality consistent.

tldr: I have a similar function, but it gets multiple years. I can modify it to use this func and we will be chilling imo

File: tpp-app/src/services/CalendarService.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    // https://github.com/uoftblueprint/the-period-purse/pull/95#discussion_r813506726
+    // Essentially retrieve the calendar for a given year and convert each day's data into a Symptom object for better front-end handling
+
+    // Used for testing
```
Comment (kenneth-miura) [inline tpp-app/src/services/CalendarService.js]: reminder to delete this once u get the ok for merge

File: tpp-app/src/services/CalendarService.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return yearData;
+    }
+
+    return null;
+}
+
+/**
+ * Saves user's current selected filter, selected month, and selected year, to preserve the thing they are looking at.
+ * Meant to be used when user exits calendar page in any way.
+ * @param selectedView: string
```
Comment (kenneth-miura) [inline tpp-app/src/services/CalendarService.js]: see earlier comment on javadoc

File: tpp-app/src/services/CalendarService.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+ * @param selectedMonth: number
+ * @param selectedYear: number
+ */
+
+export const POSTMostRecentCalendarState = async (selectedView, selectedMonth, selectedYear) => new Promise( async (resolve, reject) => {
```
Comment (kenneth-miura) [inline tpp-app/src/services/CalendarService.js:36]: what a beauty

**Review by kenneth-miura — CHANGES_REQUESTED** [review]
Looks good! just update some comments, besides that looks great! must be nice to have that youthful sharp wit! couldn't be me! :):):):):):):):):):):)

but fr tho nice pr just some small things

File: tpp-app/src/services/CalendarService.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+})
+
+/**
+ * Retrieves user's previous selected filter, selected month, and selected year, to preserve the thing they are looking at.
+ * Meant to be used when user returns to the calendar page in any way.
+ * @returns: list[selectedView, selectedMonth, selectedYear]
+ */
+export const GETMostRecentCalendarState = async () => {
+
+    var tracking = await GETAllTrackingPreferences()
```
Comment (faye-tan) [inline tpp-app/src/services/CalendarService.js]: Edward, maybe we merge this PR in after @annie951977 's since you need to import her GETAllTrackingPreferences from Settings

File: tpp-app/src/services/CalendarService.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+export const POSTMostRecentCalendarState = async (selectedView, selectedMonth, selectedYear) => new Promise( async (resolve, reject) => {
+
+    // Check if it's a valid view or month
+    if (selectedMonth > 12 || selectedMonth < 1) {
+        reject("No month to record")
+    }
+
+    var exists = Object.keys(TRACK_SYMPTOMS).some(function(k) {
+        return TRACK_SYMPTOMS[k] === selectedView;
```
Comment (yuemily24) [inline tpp-app/src/services/CalendarService.js]: 1. Small note but try and avoid using `var` because it usually implies a global variable. It's actually fine here because it's functionally scoped, but overall in JavaScript it's good practice to stick with `let` or `const` unless you want it to be globally-scoped, in which case we use `var`. Ref: [MDN doc for var](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var)

2. Since we're sticking with arrow notation I recommend the following, where `tracking` is equivalent to `TRACK_SYMPTOMS[k]`.

```suggestion
    const exists = Object.keys(TRACK_SYMPTOMS).some((tracking) => tracking === selectedView);
```

File: tpp-app/src/services/CalendarService.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    // This is used for testing
+    // TODO: This is HARDCODED
+    // tracking = {
+    //     flow: true, 
+    //     mood: true, 
+    //     sleep: true, 
+    //     cramps: true, 
+    //     exercise: true
+    // }
```
Comment (yuemily24) [inline tpp-app/src/services/CalendarService.js]: Remove comment? Unsure because there's a TODO in it, in which if you need it for the future ignore this

File: tpp-app/src/services/CalendarService.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        const selectedView = await AsyncStorage.getItem(KEYS.SELECTED_VIEW);
+        const selectedMonth = await AsyncStorage.getItem(KEYS.SELECTED_MONTH);
+        const selectedYear = await AsyncStorage.getItem(KEYS.SELECTED_YEAR);
+        
+        // Checks which view it is on
+        // TODO: this is HARDCODED
+        let index = Object.values(TRACK_SYMPTOMS).indexOf(selectedView)        
+        
+        if (tracking[Object.keys(tracking)[index]]) {
+            console.log("DEBUG: " + selectedMonth)
```
Comment (yuemily24) [inline tpp-app/src/services/CalendarService.js]: Also not sure if this was originally only for debugging?

**Review by yuemily24 — COMMENTED** [review]
Some very small comments, half of which are wondering if you forgot to delete some comments lol. Looks amazing :)

File: tpp-app/src/services/CalendarService.js (hunk 1 above)
Comment (edwardhuahan) [reply tpp-app/src/services/CalendarService.js:7]: ur old

File: tpp-app/src/services/CalendarService.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import { KEYS, TRACK_SYMPTOMS } from './utils/constants.js'
+import { Symptoms } from './utils/models.js';
+
```
Comment (faye-tan) [inline tpp-app/src/services/CalendarService.js:5]:
```suggestion
import { GETAllTrackingPreferences } from './SettingsService.js'; 

```
should be in line with @annie951977's code. Just add that and I think its ok if u merge before her

**Review by kenneth-miura — APPROVED** [review]
lgtm!

**Review by yuemily24 — APPROVED** [review]
Looks good!

**Review by annie951977 — APPROVED** [review]
Nice!
